includes/woocommerce-memberships/class-subscriptions-integration.php

```diff
@@ -0,0 +1,194 @@
+<?php
+/**
+ * Newspack Network WooCommerce Subscriptions integration for WooCommerce Memberships.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack_Network\Woocommerce_Memberships;
+
+use Newspack_Network\Incoming_Events\Membership_Plan_Updated;
+use Newspack_Network\Incoming_Events\Subscription_Changed;
+
+/**
+ * Handles tweaks for WooCommerce Memberships WooCommerce Subscriptions integration.
+ *
+ * This class handles a very specific case when a local User Membership is linked to a Subscription in this site, and then the Subscription is cancelled,
+ * triggering a cancellation to that membership.
+ *
+ * This class will short circuit the cancellation process and check if the user has an active subscription in another site that is linked to a Membership plan with the same Network ID.
+ */
+class Subscriptions_Integration {
+
+	/**
+	 * Runs the initialization.
+	 *
+	 * @return void
+	 */
+	public static function init() {
+		// You'd think this first filter is enough, but it's not. Even if the membership cancellation via linked subscription
+		// is prevented, the expiry code is also executed.
+		add_filter( 'wc_memberships_cancel_subscription_linked_membership', [ __CLASS__, 'prevent_membership_expiration' ], 10, 2 );
+		add_filter( 'wc_memberships_expire_user_membership', [ __CLASS__, 'prevent_membership_expiration' ], 10, 2 );
+	}
+
+	/**
+	 * Prevent membership expiration if another network site has a synced membership active.
+	 *
+	 * @param bool                                                      $cancel_or_expire whether to cancel/expire the membership when the subscription is cancelled (default true).
+	 * @param \WC_Memberships_Integration_Subscriptions_User_Membership $user_membership the subscription-tied membership.
+	 */
+	public static function prevent_membership_expiration( $cancel_or_expire, $user_membership ) {
+		$plan_network_id = get_post_meta( $user_membership->get_plan()->get_id(), Admin::NETWORK_ID_META_KEY, true );
+
+		if ( ! $plan_network_id ) {
+			return $cancel_or_expire;
+		}
+
+		$is_managed = get_post_meta( $user_membership->get_id(), Admin::NETWORK_MANAGED_META_KEY, true );
+
+		if ( $is_managed ) {
+			return false;
+		}
+
+		$user = $user_membership->get_user();
+		if ( ! $user ) {
+			return $cancel_or_expire;
+		}
+
+		$user_id = $user->ID;
+
+		$user_network_subs = self::get_user_network_active_subscriptions( $user_id );
+
+		$network_active_subscription = self::user_has_active_subscription_in_network( $user_id, $plan_network_id );
+		if ( $network_active_subscription ) {
+			$user_membership->add_note(
```

Comment:
In my testing, this note was added five times. Suggestion: store the timestamp of adding a note in post meta of the membership, and only add if not added in the last e.g. 5 minutes.

Reply:
https://github.com/Automattic/newspack-network/pull/135/commits/203ea707e80d33deef50fa771a387932b723646c